capsules/src/virtual_hmac.rs

```diff
@@ -0,0 +1,104 @@
+//! Virtualize the HMAC interface to enable multiple users of an underlying
+//! HMAC hardware peripheral.
+
+use kernel::common::cells::OptionalCell;
+use kernel::common::{List, ListLink, ListNode};
+use kernel::hil::digest;
+use kernel::hil::digest::DigestType;
+use kernel::ReturnCode;
+
+pub struct VirtualMuxHmac<
+    'a,
+    A: kernel::hil::digest::HMACSha256 + digest::Digest<'a, T>,
+    T: DigestType,
+> {
+    mux: &'a MuxHmac<'a, A, T>,
+    next: ListLink<'a, VirtualMuxHmac<'a, A, T>>,
+    client: OptionalCell<&'a dyn digest::Client<'a>>,
+}
+
+impl<A: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType>
+    ListNode<'a, VirtualMuxHmac<'a, A, T>> for VirtualMuxHmac<'a, A, T>
+{
+    fn next(&self) -> &'a ListLink<VirtualMuxHmac<'a, A, T>> {
+        &self.next
+    }
+}
+
+impl<A: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> VirtualMuxHmac<'a, A, T> {
+    pub fn new(mux_alarm: &'a MuxHmac<'a, A, T>) -> VirtualMuxHmac<'a, A, T> {
+        VirtualMuxHmac {
+            mux: mux_alarm,
+            next: ListLink::empty(),
+            client: OptionalCell::empty(),
+        }
+    }
+}
+
+impl<A: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> digest::Digest<'a, T>
+    for VirtualMuxHmac<'a, A, T>
+{
+    /// Set the client instance which will receive `add_data_done()` and
+    /// `hash_done()` callbacks
+    fn set_client(&'a self, client: &'a dyn digest::Client<'a>) {
+        self.mux.virtual_hmacs.push_head(self);
+        self.mux.hmac.set_client(client);
+    }
+
+    /// Add data to the hmac IP.
+    /// All data passed in is fed to the HMAC hardware block.
+    /// Returns the number of bytes written on success
+    fn add_data(&self, data: &'static mut [u8]) -> Result<usize, (ReturnCode, &'static mut [u8])> {
+        self.mux.hmac.add_data(data)
+    }
```

Comment:
I don't think this actually virtualizes, does it? Wouldn't two `VirtualMuxHmac`s using this interface at the same time interleave their data?

Reply:
Yep :(

Reply:
I had a go at actually muxing this. I can't think of a good way to mux the key selection. Any ideas?

Reply:
I added a basic only one user at a time check. I couldn't think of a nice way to support multiple users at the same time as we would need to buffer data which I was trying to avoid.